docs/topics/solidity/guideline.md

```diff
@@ -0,0 +1,293 @@
+
+```
+// Reserved storage space to allow for layout changes in the future.
+uint256[50] private ______gap;
+```
+
+### Use a separate contract for storage
+
+Create one storage contract and then inherit it into the implementation. This way it is easier to manage the storage and
+add variables during upgrade.
+
+### Loop length must be short and controlled
+
+All the iterations in the smart contract must be limited in their length. A permanent denial of service could happen if
+the gas consumed by a loop is too high.
+
+### Rounding
+
+Each time a division is done the numbers are rounded (the operation is like a `floor(a/b)`) and there is some remainder
+to take into account.
+
+## Visibility
+
+### TO DISCUSS: Use of private attributes
+
+Private attributes just hide the internal state of the contract, child contracts (via inheritance), and make writing
+tests impossible.
+
+***When possible use internal or public.***
+
+Internal solve the testing issue. A mock child contract can be used to access the variable.
+
+Public solve all the issues the only downside is that it adds a default getter and calling a public method can make the
+contract code and gas consumption higher.
+
+### TO DISCUSS: Use of private view methods
+
+Usually view methods are used as a getters or to do some reusable calculation. Exposing the method helps to test the
+contract and is useful to check the internal state of a deployed contract. Methods can be exposed using public or
+external. If possible prefer external because external function's parameters are not copied into memory but are read
+directly from `calldata` and this can save some gas.
+
+### Calling internal functions is cheaper
+
+Calling internal functions is cheaper than calling public functions. When you call a public function all the parameters
+are copied again into memory and passed to that function. By contrast, when you call an internal function, references of
+those parameters are passed, and they are not copied again.
+
+## Others
+
+### Lock pragmas to specific compiler version
+
+Pragma must select a specific version of the compiler, instead of using `pragma solidity ^0.8.2;`
+use `pragma solidity 0.8.2;`.
+
+### Use interface type instead of the address for type safety
+
+```
+interface SomethingToCall {
+    function call() returns (bool);
+}
+
+contract Example {
+    function Good(SomethingToCall _something) internal returns (bool) {
+    }
+
+    function Bad(address _addr) internal returns (bool) {
+        SomethingToCall something = SomethingToCall(_addr);
+    }
+}
+```
+
+### Use short reason strings
+
+The solidity linter warns about it. Reason strings (those used in `revert`) that takes more than 32 bytes make the
+contract byte code bigger.
+
+### TO DISCUSS: Pull over Push
+
+From the list of [Solidity Patterns](https://github.com/fravoll/solidity-patterns) this one is a very important one: "
+Sending ether to another address in Ethereum involves a call to the receiving entity".
+
```

Comment:
From my understanding, this pattern can be useful but it also have downsides: one more transaction,etc

https://github.com/fravoll/solidity-patterns/blob/master/docs/pull_over_push.md#consequences

Not really a best practice but a pattern that can be useful

Reply:
My doubt here is if this pattern is not a must when you send funds to an address that is not under your control ?

Reply:
Not sure i understand your question but here the 3 reasons to use it 

```
Use the Pull over Push pattern when

- you want to handle multiple ether transfers with one function call.
- you want to avoid taking the risk associated with ether transfers.
- there is an incentive for your users to handle ether withdrawal on their own.
```

Reply:
Ok, I replaced this part with something about "Sending funds to an uncontrolled address"